File: dotnet/src/MableBank.Infrastructure/InputFile.cs
namespace MableBank.Infrastructure;

public sealed class InputFileValidationException : Exception
{
    public InputFileValidationException()
    {
    }

    public InputFileValidationException(string message)
        : base(message)
    {
    }

    public InputFileValidationException(string message, Exception innerException)
        : base(message, innerException)
    {
    }
}

public static class InputFile
{
    private const long MaxBytes = 10 * 1024 * 1024;

    public static string Validate(string? path)
    {
        if (string.IsNullOrWhiteSpace(path))
        {
            throw new InputFileValidationException("Path is required");
        }

        var expanded = Path.GetFullPath(path);
        var fileInfo = new FileInfo(expanded);
        if (!fileInfo.Exists)
        {
            throw new InputFileValidationException($"File not found: {path}");
        }

        if (fileInfo.LinkTarget is not null)
        {
            throw new InputFileValidationException($"Not a regular file: {path}");
        }

        if (fileInfo.Length > MaxBytes)
        {
            throw new InputFileValidationException(
                $"File too large (max {MaxBytes} bytes): {path}");
        }

        return expanded;
    }
}

File: ruby/lib/mable_bank/infrastructure/input_file.rb
# frozen_string_literal: true

module MableBank
  module Infrastructure
    class InputFile
      class ValidationError < StandardError; end

      MAX_BYTES = 10 * 1024 * 1024

      def self.validate!(path)
        raise ValidationError, 'Path is required' if path.nil? || path.to_s.strip.empty?

        expanded = File.expand_path(path)
        raise ValidationError, "File not found: #{path}" unless File.file?(expanded)
        raise ValidationError, "Not a regular file: #{path}" if File.symlink?(expanded)
        raise ValidationError, "File too large (max #{MAX_BYTES} bytes): #{path}" if File.size(expanded) > MAX_BYTES

        expanded
      end
    end
  end
end
